README.md
        ____  __                        __    __         _
       / __ \/ /_  __ ___   ___  ____ _/ /_  / /__      (_)____
      / /_/ / / / / / __ \/ __ \/ __/ / __ \/ / _ \    / / ___/
     / ____/ / /_/ / /_/ / /_/ / /_/ / /_/ / /  __/   / (__  )
    /_/   /_/\__,_/\__, /\__, /\__/_/_.___/_/\___(_)_/ /____/
                  /____//____/                    /___/

[![Travis](https://api.travis-ci.org/conversejs/pluggable.js.png?branch=master)](https://travis-ci.org/jcbrand/pluggable.js)

# Introduction

pluggable.js is a tiny library that lets you make your JS project extendable via plugins,
while still keeping sensitive objects and data private through closures.

It was originally written for [converse.js](https://conversejs.org), to provide
a plugin architecture that allows 3rd party developers to extend and override
private objects and classes, but it does not require nor depend on either library.

Pluggable.js is only 2.5KB when minified and uncompressed.

# Documentation

To understand how it works under the hood, read the [annotated source code](
https://jcbrand.github.io/pluggable.js/docs/pluggable.html).

## Live demo of the example code

You can try out the [live demo](https://jcbrand.github.io/pluggable.js/example)
of the example code given below.

The files themselves are in the [example folder of the
repo](https://github.com/jcbrand/pluggable.js/tree/master/example).

## Running the tests

Simply run `make check`

## Usage

*Please note: The examples to follow use the ES5 version of Javascript.
 Pluggable.js however also works with ES2015 modules.*

Suppose you have a module with a private scope and two private methods,
`_showNotification` and `_fadeOut`. Let's just assume for the sake of
illustration that there's a good reason why these methods are private.

What we want to do, is to make this module pluggable. In other words, we want
people to be able to write plugins for this module, in which they can:

- Access the private or closured scope of this module.
- Add new methods to the module.
- Override or extend existing methods (private or public) on this module.

To make the module pluggable, we simply call `pluggable.enable(module);`.
Once we've made this call, the `module` object will have a new property
`pluginSocket`, which you can think of as the thing into which the plugins are
plugged into. It is an instance of `PluginSocket`, which represents the plugin
architecture and manages all the plugins.

Additionally, we need to expose some way for plugins to register themselves.
Since the module itself is private, we'll need to expose the `registerPlugin`
method on the `pluginSocket` via a public method.

And then finally, once all plugins have been registered, they need to be
initialized by calling `initializePlugins` on the `PluginSocket` instance.

For example:

``` Javascript
    module.pluginSocket.initializePlugins();
```

### Example code of a module to be made pluggable

Let's now look at the code for our module (which will go into a file called `app.js`):

``` Javascript
// An example application, used to demonstrate how pluggable.js can
// allow a module to be made pluggable.
(function () {
    var module = this;

    // A private method, not available outside the scope of this module.
    module._showNotification = function (text) {
        /* Displays a notification message */
        var el = document.createElement('p');
        el.setAttribute('class', 'notification');
        el.appendChild(document.createTextNode(text));
        el = document.getElementById('notifications').appendChild(el);
        setTimeout(_.partial(module._fadeOut, el), 3000);
        return el;
    };

    // Another private method, not available outside the scope of this module.
    module._fadeOut = function (el) {
        /* Fades out the passed in DOM element. */
        el.style.opacity = 1;
        (function fade() {
            if ((el.style.opacity -= 0.1) < 0) {
                el.remove();
            } else {
                setTimeout(fade, 25);
            }
        })();
    };

    // Initialize this module.
    // -----------------------
    // There are two buttons for which we register event handlers.
    //
    // This will be a public method.
    module.initialize = function () {
        var notify_el = document.getElementById('notify');
        notify_el.addEventListener('click', function () {
            module._showNotification('This is a notification.');
        });

        var enable_el = document.getElementById('enable');
        enable_el.addEventListener('click', function () {
            this.setAttribute('disabled', 'disabled');
            module.pluginSocket.initializePlugins();
        });
    };

    // Register a plugin for this module.
    // ----------------------------------
    // This is a wrapper method which defers to the `registerPlugin` method
    // of pluggable.js, which is on the `pluginSocket` property of the
    // private `module` object that was made pluggable, via
    // `pluggable.enable(module).
    //
    // This will be a public method.
    module.registerPlugin = function (name, plugin) {
        module.pluginSocket.registerPlugin(name, plugin);
    };

    // Calling `pluggable.enable` on the closured `module` object, will make it
    // pluggable. Additionally, it will get the `pluginSocket` attribute, which
    // refers to the object that the plugins get plugged into.
    pluggable.enable(module);

    // Declare the two public methods
    var _public = {
        'initialize': module.initialize,
        'registerPlugin': module.registerPlugin
    };
    window.myApp = _public;
    return _public;
})();
```

### Creating the plugin

So, as you can see in the example above, the module has a private method
`_showNotification`, which will show a notification message in the page.

Let's now create a plugin which overrides this method, so that we can modify the
way the notification message will be displayed.

By the way, multiple plugins can override the same method. As long as
each overriding method calls the original method, via the `__super__` property,
the method call will travel up through all the overrides (in the reverse order
in which the plugins were registered) back to the original method.

The way to call the method being overridden, is like this:

``` Javascript
    this.__super__.methodName.apply(this, arguments);
```

where `methodName` is the name of the method. In our example, this will be
`showNotification`.

### Example plugin code

So, with that said, here's what the plugin code (which goes into the file `plugin.js`)
will look like:

``` Javascript
(function () {
    // We call the public `registerPlugin` method from our module in app.js
    window.myApp.registerPlugin('my-plugin', {

        overrides: {
            // Here you specify your overrides of methods or objects on the
            // module that has been made pluggable.

            // Override _showNotification to change the color of notification
            // messages.
            _showNotification: function (text) {
                // When overriding a method, you can still call the original method
                // via `this.__super__`. To properly call with the proper `this`
                // context and parameters, use `.apply(this, arguments)`.
                var el = this.__super__._showNotification.apply(this, arguments);

                // Now we simply set another class on the element containing
                // the notifications, so that they'll appear in a different
                // color.
                el.setAttribute('class', el.getAttribute('class')+' extra');
            }

            // BTW, new methods which don't exist yet can also be added here.
        },

        initialize: function (socket) {
            // The initialize function gets called as soon as the plugin is
            // loaded by pluggable.js's plugin machinery.

            // We are passed in an instance of the  `PluginSocket`, which
            // represents the plugin architecture and gets created when
            // `pluggable.enable` is called on an object.

            // The `PluginSocket` instance is also accessible via the `pluginSocket`
            // attribute, on the module that was passed to `pluggable.enable`.

            // You can get hold of the module which was made pluggable
            // via the `plugged` property of the `PluginSocket` instance.

            // Once you have the module, you can access its private properties
            // and call its private methods.

            // Here, once this plugin is initialized, we show a notification.
            socket.plugged._showNotification(
                "The plugin has been enabled. Notifications are now a different color."
            );
        }
    });
})();
```
### Custom plugin code

The plugin has three important parts to it. Firstly, a call to register the
plugin must be made, and then passed in with this call is an object which
represents the plugin itself.

This plugin object has an `overrides` key, which is another object containing methods
and objects that will override similarly named methods and object properties on the
module itself.

Additionally, the `initializePlugins` method has been called on the
`pluginSocket` instance.

This happens inside our module in app.js and this methods should generally be
called once, after all plugins have been registered and after the module being
made pluggable has itself been initialized.

In the `initialize` method, you have access the module's scope, which would
otherwise not be available to you.

So, as you can see, the plugin lets us achieve our three goals stated earlier:

### Access the closured scope of this module.

We have access to the module and all its properties and methods inside the
`initialize` method of our plugin.

### Add new methods to the module.

We can add new methods to our module, either inside the `initialize` method of
our plugin, or by stating them declaratively in `overrides` object.

### Override or extend existng methods (private or public) on this module.

We can override existing methods and object properties via the `overrides`
object.

# Contact

For issues and requests, please use the project's [issue tracker](https://github.com/jcbrand/pluggable.js/issues).

You can reach me via my website's [contact form](http://opkode.com/contact.html) and you can also [follow me on twitter](https://twitter.com/jcopkode).


CHANGES.md
# Changelog

## 3.0.1 (2021-05-12)

- Bugfix. Plugin `dependencies` attributes might not be defined
- Bugfix. `this` value is wrong in overridden functions

## 3.0.0 (2021-05-12)

- Remove lodash and use native methods instead

## 2.0.1 (2019-05-13)

- Import lodash methods individually
- Log not-found dependency messages as warnings.

## 2.0.0 (2018-01-10)

- Rename the `optional_dependencies` option to `dependencies`, since they're
  not necessarily optional and the original name is misleading.

## 1.0.1 (2017-11-05)

- Ensure that `plugged` obj is also available in overriden constructors
- Plugins now can have an `enabled` attribute (either a method or a property)
  which determines whether the plugin should be initialized or not.

## 1.0.0 (2017-02-25)

- Use lodash instead of underscore.
- Update pluggable.js to ES2015 syntax.
- Add unit tests and integrate into TravisCI.
- Throw an error when reusing an already taken plugin name.

## 0.0.3 (15 January 2017)
- Ensure that pluggable obj is reachable in overridden functions

## 0.0.2 (24 August 2016)
- Rename `_super` attribute to `__super__`.
- Renamed `Pluggable` class to `PluginSocket`.
- Allow the attribute name referring to the `PluginSocket` instance to be
  specified when calling `pluggable.enable`.

## 0.0.1
- Initial release


tests.js
import test from 'tape';
import * as _ from 'lodash';
import * as pluggable from './pluggable';

test('A passing test', (assert) => {
  assert.pass('This test will pass.');
  assert.end();
});

test('Assertions with tape.', (assert) => {
  const expected = 'something to test';
  const actual = 'sonething to test';
  assert.equal(actual, expected,
    'Given two mismatched values, .equal() should produce a nice bug report');
  assert.end();
});


tests/app.js
const app = {};

import pluggable from '../dist/pluggable.js';

function initialize (whitelist=[], blacklist=[]) {
    app.pluginSocket.initializePlugins({}, whitelist, blacklist);
}

function registerPlugin (name, plugin) {
    app.pluginSocket.registerPlugin(name, plugin);
}

function getPluginSocket () {
    // Normally this wouldn't be exposed, but we do so here for testing
    // purposes.
    return app.pluginSocket;
}

function getClosuredApp () {
    // Normally this wouldn't be exposed, but we do so here for testing
    // purposes.
    return app;
}

// Calling `pluggable.enable` on the private `app` object, will make it
// pluggable. Additionally, it will get the `pluginSocket` attribute, which
// refers to the object that the plugins get plugged into.
function makePluggable () {
    pluggable.enable(app);
}

export {
    initialize,
    registerPlugin,
    getClosuredApp,
    getPluginSocket,
    makePluggable
};


tests/tests.js
import * as app from './app';
import plugin from './plugin';
import test from 'tape';


test('Plugin registration', (assert) => {
    assert.ok(
        app.getPluginSocket() === undefined,
        "Non-pluggable modules don't have the 'pluginSocket' attribute"
    );
    app.makePluggable();
    assert.ok(
        app.getPluginSocket() !== undefined,
        "Pluggable modules have the 'pluginSocket' attribute"
    );

    assert.ok(
        Object.keys(app.getPluginSocket().plugins).length === 0,
        "By default there aren't any registerd pugins"
    );
    app.registerPlugin('my-plugin', plugin);
    assert.ok(
        Object.keys(app.getPluginSocket().plugins).length === 1,
        'A plugin is registered by calling registerPlugin'
    );
    assert.ok(
        app.getPluginSocket().plugins['my-plugin'] === plugin,
        'The object passed in to registerPlugin is saved as a plugin'
    );
    assert.throws(
        () => app.registerPlugin('my-plugin', {}),
        Error,
        'Registering a plugin under an already taken name results in an Error exception'
    );
    assert.end();
});

test('disable initialization', (assert) => {
    app.makePluggable();
    let disabled_plugin_initialized = false;
    let enabled_plugin_initialized = false;

    app.registerPlugin('disabled-plugin', {
        enabled () { return false },
        overrides: {
            foo: 'bar'
        },
        initialize () {
            disabled_plugin_initialized = true;
        }
    });
    app.registerPlugin('enabled-plugin', {
        enabled () { return true },
        overrides: {
            baz () { return 'buz' }
        },
        initialize () {
            enabled_plugin_initialized = true;
        }
    });
    app.initialize();

    assert.ok(
        enabled_plugin_initialized === true &&
        disabled_plugin_initialized === false,
        "A plugin with an 'enable' method which returns false, is not initialized"
    )
    assert.ok(
        app.getClosuredApp().baz() === 'buz' &&  ('foo' in app.getClosuredApp()) === false,
        "A plugin with an 'enable' method which returns false, does not have its overrides applied"
    )
    assert.end();
});

test('blacklisting of plugins', (assert) => {
    app.makePluggable();
    app.registerPlugin('allowed-plugin', {});
    app.registerPlugin('blacklisted-plugin', {});
    app.initialize([], ['blacklisted-plugin']);

    const allowed_plugins = app.getPluginSocket().allowed_plugins;
    assert.ok(
        ('blacklisted-plugin' in allowed_plugins) === false &&
        ('allowed-plugin' in allowed_plugins) === true,
        "A blacklisted plugin will be excluded from the allowed_plugins list"
    );

    const initialized_plugins = app.getPluginSocket().initialized_plugins;
    assert.ok(
        initialized_plugins.includes('blacklisted-plugin') === false &&
        initialized_plugins.includes('allowed-plugin') === true,
        "A blacklisted plugin won't be initialized"
    );
    assert.end();
});

test('whitelisting of plugins', (assert) => {
    app.makePluggable();
    app.registerPlugin('whitelisted-plugin', {});
    app.registerPlugin('nonlisted-plugin', {});
    app.initialize(['whitelisted-plugin'], []);

    let allowed_plugins = app.getPluginSocket().allowed_plugins;
    assert.ok(
        ('nonlisted-plugin' in allowed_plugins) === false &&
        ('whitelisted-plugin' in allowed_plugins) === true,
        "A non-whitelisted plugin will not be in the allowed_plugins list"
    );

    let initialized_plugins = app.getPluginSocket().initialized_plugins;
    assert.ok(
        initialized_plugins.includes('nonlisted-plugin') === false &&
        initialized_plugins.includes('whitelisted-plugin') === true,
        "A non-whitelisted plugin won't be initialized"
    );
    assert.end();
});

test('white- and blacklisting of plugins', (assert) => {
    app.makePluggable();
    app.registerPlugin('allowed-plugin', {});
    app.registerPlugin('blacklisted-plugin', {});
    app.initialize(['allowed-plugin', 'blacklisted-plugin'], ['blacklisted-plugin']);

    let allowed_plugins = app.getPluginSocket().allowed_plugins;
    assert.ok(
        ('blacklisted-plugin' in allowed_plugins) === false &&
        ('allowed-plugin' in allowed_plugins) === true,
        "A whitelisted plugin that's also blacklisted won't be in the allowed_plugins list"
    );

    let initialized_plugins = app.getPluginSocket().initialized_plugins;
    assert.ok(
        initialized_plugins.includes('blacklisted-plugin') === false &&
        initialized_plugins.includes('allowed-plugin') === true,
        "A whitelisted plugin that's also blacklisted won't be initialized"
    );
    assert.end();
});


src/pluggable.js
/*
       ____  __                        __    __         _
      / __ \/ /_  __ ___   ___  ____ _/ /_  / /__      (_)____
     / /_/ / / / / / __ \/ __ \/ __/ / __ \/ / _ \    / / ___/
    / ____/ / /_/ / /_/ / /_/ / /_/ / /_/ / /  __/   / (__  )
   /_/   /_/\__,_/\__, /\__, /\__/_/_.___/_/\___(_)_/ /____/
                 /____//____/                    /___/
 */

// Pluggable.js lets you to make your Javascript code pluggable while still
// keeping sensitive objects and data private through closures.


// `wrappedOverride` creates a partially applied wrapper function
// that makes sure to set the proper super method when the
// overriding method is called. This is done to enable
// chaining of plugin methods, all the way up to the
// original method.
function wrappedOverride (key, value, super_method, default_super, ...args) {
    if (typeof super_method === "function") {
        if (typeof this.__super__ === "undefined") {
            /* We're not on the context of the plugged object.
             * This can happen when the overridden method is called via
             * an event handler or when it's a constructor.
             *
             * In this case, we simply tack on the  __super__ obj.
             */
            this.__super__ = default_super;
        }
        this.__super__[key] = super_method.bind(this);
    }
    return value.apply(this, args);
}


// The `PluginSocket` class contains the plugin architecture, and gets
// created whenever `pluggable.enable(obj);` is called on the object
// that you want to make pluggable.
// You can also see it as the thing into which the plugins are plugged.
// It takes two parameters, first, the object being made pluggable, and
// then the name by which the pluggable object may be referenced on the
// __super__ object (inside overrides).
class PluginSocket {

    constructor (plugged, name) {
        this.name = name;
        this.plugged = plugged;
        if (typeof this.plugged.__super__ === 'undefined') {
            this.plugged.__super__ = {};
        } else if (typeof this.plugged.__super__ === 'string') {
            this.plugged.__super__ = { '__string__': this.plugged.__super__ };
        }
        this.plugged.__super__[name] = this.plugged;
        this.plugins = {};
        this.initialized_plugins = [];
    }

    // `_overrideAttribute` overrides an attribute on the original object
    // (the thing being plugged into).
    //
    // If the attribute being overridden is a function, then the original
    // function will still be available via the `__super__` attribute.
    //
    // If the same function is being overridden multiple times, then
    // the original function will be available at the end of a chain of
    // functions, starting from the most recent override, all the way
    // back to the original function, each being referenced by the
    // previous' __super__ attribute.
    //
    // For example:
    //
    // `plugin2.MyFunc.__super__.myFunc => plugin1.MyFunc.__super__.myFunc => original.myFunc`
    _overrideAttribute (key, plugin) {
        const value = plugin.overrides[key];
        if (typeof value === "function") {
            const default_super = {};
            default_super[this.name] = this.plugged;
            const super_method = this.plugged[key];
            this.plugged[key] = function (...args) {
                return wrappedOverride.apply(this, [key, value, super_method,  default_super, ...args]);
            }
        } else {
            this.plugged[key] = value;
        }
    }

    _extendObject (obj, attributes) {
        if (!obj.prototype.__super__) {
            obj.prototype.__super__ = {};
            obj.prototype.__super__[this.name] = this.plugged;
        }
        for (const [key, value] of Object.entries(attributes)) {
            if (key === 'events') {
                obj.prototype[key] = Object.assign(value, obj.prototype[key]);
            } else if (typeof value === 'function') {
                // We create a partially applied wrapper function, that
                // makes sure to set the proper super method when the
                // overriding method is called. This is done to enable
                // chaining of plugin methods, all the way up to the
                // original method.
                const default_super = {};
                default_super[this.name] = this.plugged;
                const super_method = obj.prototype[key];
                obj.prototype[key] = function (...args) {
                    return wrappedOverride.apply(this, [key, value, super_method, default_super, ...args]);
                }
            } else {
                obj.prototype[key] = value;
            }
        }
    }

    // Plugins can specify dependencies (by means of the
    // `dependencies` list attribute) which refers to dependencies
    // which will be initialized first, before the plugin itself gets initialized.
    //
    // If `strict_plugin_dependencies` is set to `false` (on the object being
    // made pluggable), then no error will be thrown if any of these plugins aren't
    // available.
    loadPluginDependencies (plugin) {
        plugin.dependencies?.forEach(name => {
            const dep = this.plugins[name];
            if (dep) {
                if (dep.dependencies?.includes(plugin.__name__)) {
                    /* FIXME: circular dependency checking is only one level deep. */
                    throw "Found a circular dependency between the plugins \""+
                        plugin.__name__+"\" and \""+name+"\"";
                }
                this.initializePlugin(dep);
            } else {
                this.throwUndefinedDependencyError(
                    "Could not find dependency \""+name+"\" "+
                    "for the plugin \""+plugin.__name__+"\". "+
                    "If it's needed, make sure it's loaded by require.js");
            }
        });
    }

    throwUndefinedDependencyError (msg) {
        if (this.plugged.strict_plugin_dependencies) {
            throw msg;
        } else {
            if (console.warn) {
                console.warn(msg);
            } else {
                console.log(msg);
            }
        }
    }

    // `applyOverrides` is called by initializePlugin. It applies any
    // and all overrides of methods or Backbone views and models that
    // are defined on any of the plugins.
    applyOverrides (plugin) {
        Object.keys(plugin.overrides || {}).forEach(key => {
            const override = plugin.overrides[key];
            if (typeof override === "object") {
                if (typeof this.plugged[key] === 'undefined') {
                    this.throwUndefinedDependencyError(
                        `Plugin "${plugin.__name__}" tried to override "${key}" but it's not found.`);
                } else {
                    this._extendObject(this.plugged[key], override);
                }
            } else {
                this._overrideAttribute(key, plugin);
            }
        });
    }

    // `initializePlugin` applies the overrides (if any) defined on all
    // the registered plugins and then calls the initialize method of the plugin
    initializePlugin (plugin) {
        if (!Object.keys(this.allowed_plugins).includes(plugin.__name__)) {
            /* Don't initialize disallowed plugins. */
            return;
        }
        if (this.initialized_plugins.includes(plugin.__name__)) {
            /* Don't initialize plugins twice, otherwise we get
            * infinite recursion in overridden methods.
            */
            return;
        }
        if (typeof plugin.enabled === 'boolean' && plugin.enabled ||
            plugin.enabled?.(this.plugged) ||
            plugin.enabled == null) { // isNil

            Object.assign(plugin, this.properties);
            if (plugin.dependencies) {
                this.loadPluginDependencies(plugin);
            }
            this.applyOverrides(plugin);
            if (typeof plugin.initialize === "function") {
                plugin.initialize.bind(plugin)(this);
            }
            this.initialized_plugins.push(plugin.__name__);
        }
    }

    // `registerPlugin` registers (or inserts, if you'd like) a plugin,
    // by adding it to the `plugins` map on the PluginSocket instance.
    registerPlugin (name, plugin) {
        if (name in this.plugins) {
            throw new Error('Error: Plugin name '+name+' is already taken');
        }
        plugin.__name__ = name;
        this.plugins[name] = plugin;
    }

    // `initializePlugins` should get called once all plugins have been
    // registered. It will then iterate through all the plugins, calling
    // `initializePlugin` for each.
    // The passed in  properties variable is an object with attributes and methods
    // which will be attached to the plugins.
    initializePlugins (properties={}, whitelist=[], blacklist=[]) {
        if (!Object.keys(this.plugins).length) {
            return;
        }
        this.properties = properties;
        this.allowed_plugins = {};

        for (const [key, plugin] of Object.entries(this.plugins)) {
            if ((!whitelist.length || whitelist.includes(key)) && !blacklist.includes(key)) {
                this.allowed_plugins[key] = plugin;
            }
        }
        Object.values(this.allowed_plugins).forEach(o => this.initializePlugin(o));
    }
}

function enable (object, name, attrname) {
    // Call the `enable` method to make an object pluggable
    //
    // It takes three parameters:
    // - `object`: The object that gets made pluggable.
    // - `name`: The string name by which the now pluggable object
    //     may be referenced on the __super__ obj (in overrides).
    //     The default value is "plugged".
    // - `attrname`: The string name of the attribute on the now
    //     pluggable object, which refers to the PluginSocket instance
    //     that gets created.
    if (typeof attrname === "undefined") {
        attrname = "pluginSocket";
    }
    if (typeof name === 'undefined') {
        name = 'plugged';
    }
    object[attrname] = new PluginSocket(object, name);
    return object;
}

export {
    enable
};
export default {
    enable
};


example/plugin.js
(function () {
    // We call the public `registerPlugin` method from our module in app.js
    window.myApp.registerPlugin('my-plugin', {
        overrides: {
            // Here you specify your overrides of methods or objects on the
            // module that has been made pluggable.

            // Override _showNotification to change the color of notification
            // messages.
            _showNotification: function (text) {
                // When overriding a method, you can still call the original method
                // via `this.__super__`. To properly call with the proper `this`
                // context and parameters, use `.apply(this, arguments)`.
                var el = this.__super__._showNotification.apply(this, arguments);
                el.setAttribute('class', el.getAttribute('class')+' extra');
            }

            // BTW, new functions which don't exist yet can also be added.
        },

        initialize: function (socket) {
            // The initialize function gets called as soon as the plugin is
            // loaded by pluggable.js's plugin machinery.

            // We are passed in an instance of the  `PluginSocket`, which
            // represents the plugin architecture and gets created when
            // `pluggable.enable` is called on an object.

            // The `PluginSocket` instance is also accessible via the `pluginSocket`
            // attribute, on the module that was passed to `pluggable.enable`.

            // You can get hold of the module which was made pluggable
            // via the `plugged` property of the `PluginSocket` instance.

            // Once you have the module, you can access its private properties
            // and call its private methods.

            // Here, once this plugin is initialized, we show a notification.
            socket.plugged._showNotification(
                "The plugin has been enabled. Notifications are now a different color."
            );
        }
    });
})();


example/app.js
// An example application, used to demonstrate how pluggable.js can
// allow a module to be made pluggable.
(function () {
    var module = {};

    // A private method, not available outside the scope of this module.
    module._showNotification = function (text) {
        /* Displays a notification message */
        var el = document.createElement('p');
        el.setAttribute('class', 'notification');
        el.appendChild(document.createTextNode(text));
        el = document.getElementById('notifications').appendChild(el);
        setTimeout(() => module._fadeOut(el), 3000);
        return el;
    };

    // Another private method, not available outside the scope of this module.
    module._fadeOut = function (el) {
        /* Fades out the passed in DOM element. */
        el.style.opacity = 1;
        (function fade() {
            if ((el.style.opacity -= 0.1) < 0) {
                el.remove();
            } else {
                setTimeout(fade, 25);
            }
        })();
    };

    // Initialize this module.
    // -----------------------
    // There are two buttons for which we register event handlers.
    //
    // This will be a public method.
    module.initialize = function () {
        var notify_el = document.getElementById('notify');
        notify_el.addEventListener('click', function () {
            module._showNotification('This is a notification.');
        });

        var enable_el = document.getElementById('enable');
        enable_el.addEventListener('click', function () {
            this.setAttribute('disabled', 'disabled');
            module.pluginSocket.initializePlugins();
        });
    };

    // Register a plugin for this module.
    // ----------------------------------
    // This is a wrapper method which defers to the `registerPlugin` method
    // of pluggable.js, which is on the `pluginSocket` property of the
    // private `module` object that was made pluggable, via
    // `pluggable.enable(module).
    //
    // This will be a public method.
    module.registerPlugin = function (name, plugin) {
        module.pluginSocket.registerPlugin(name, plugin);
    };

    // Calling `pluggable.enable` on the closured `module` object, will make it
    // pluggable. Additionally, it will get the `pluginSocket` attribute, which
    // refers to the object that the plugins get plugged into.
    pluggable.enable(module, 'app');

    // Declare the two public methods
    var _public = {
        'initialize': module.initialize,
        'registerPlugin': module.registerPlugin
    };
    window.myApp = _public;
    return _public;
})();
